[CDTOOL-1108] - Support for Virtual Patches in Terraform

rcaril opened the pull request:
All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/fastly/go-fastly/pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

* [x] Does your submission pass tests? 
* [x] Post the output of your test runs
<img width="1839" height="954" alt="image" src="https://github.com/user-attachments/assets/3583a50b-92fa-45e9-b5ea-297103bde743" />

`make test` output:
<img width="1436" height="320" alt="image" src="https://github.com/user-attachments/assets/46ccb72a-7daf-48b9-b794-df9f55b15791" />

`make testacc TESTARGS='-run=TestAccFastlyNGWAFVirtualPatch*'` output
<img width="1796" height="973" alt="image" src="https://github.com/user-attachments/assets/b1454371-b7db-43c1-9349-82ab50ee2ccf" />

rcaril committed:
Started changes

Work in progress

rcaril committed:
final updates

rcaril committed:
incl tests

rcaril committed:
Merge branch 'main' into CDTOOL-1108

rcaril committed:
Pre-Doc Generation

rcaril committed:
Corrected Plural Syntax of Resource + Docs

rcaril committed:
Corrected Testing Resource Syntax as well for Plural Changes

rcaril committed:
Ensured that we are using SetId for updates and imports

rcaril committed:
Updated State File ID refs to only use the primary resource ID instead of what we are using for import

<VPid> instead of <WSid/VPid>

anthony-gomez-fastly commented on fastly/resource_fastly_ngwaf_virtualpatches.go:
```diff
@@ -0,0 +1,157 @@
+
+func resourceFastlyNGWAFVirtualPatches() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceFastlyNGWAFVirtualPatchUpdate,
```
I'm not super terraform fluent, but is this how we're supposed to handle objects that can't be created via the API?

rcaril replied:
This is what @philippschulte suggested I do when I chatted with him earlier in the week. He said the Create / Delete functions still need to exist as this is what TF expects to utilize with the built in functions.

kpfleming replied:
I would expect there to be a `Create` function which returns an error of some type since the resource cannot be created by HCL. I don't know if there's an example of that already in our provider, though.

kpfleming commented on fastly/resource_fastly_ngwaf_virtualpatch_test.go:
None of these need to be checked here; checking that behavior is the responsibility of the `workspace` resource, not this one.

kpfleming commented on fastly/resource_fastly_ngwaf_virtualpatch_test.go:
This 'workspace' configuration should only include required attributes and any attributes which may be required for testing the Virtual Patch resource. The rest can be left at their default values.

kpfleming commented on fastly/resource_fastly_ngwaf_virtualpatch_test.go:
This should be identical to the `workspace` resource when it was created, there is no need to modify the workspace in this test.

kpfleming commented on fastly/resource_fastly_ngwaf_virtualpatches.go:
```diff
@@ -0,0 +1,157 @@
+			"action": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "User-configured action of the virtual patch. Accepted values are [`block`, `log`]",
```
All of the attribute descriptions should match the ones at https://branch-apidocs-console-1533-ngwaf-functional-parity.developer.fastly.com/documentation/reference/api/ngwaf/virtual-patches/ (also in the data source).

kpfleming commented on fastly/resource_fastly_ngwaf_virtualpatches.go:
```diff
@@ -0,0 +1,157 @@
+
+	log.Printf("[DEBUG] REFRESH: NGWAF virtual patch input: %#v", i)
+
+	virtualpatch, err := ws.Get(gofastly.NewContextForResourceID(ctx, d.Id()), conn, &i)
```
The ID passed to `NewContextForResourceID` must be the workspace ID, not the ID of the Virtual Patch resource.

kpfleming reviewed (changes requested):
Please also add a documentation template for the data source; you can use the one for redactions as an example.

rcaril committed:
Updates

- Corrected Params Workspace Updates (only required)
- Added Docs for Data Source
- Corrected Data source logic (was actually broken)
- Corrected descriptions
- Added custom error to handle users trying to create a new VP

rcaril commented:
Custom Create error handling: 
<img width="1527" height="211" alt="image" src="https://github.com/user-attachments/assets/a3315c3c-aee9-4b7e-a7e5-eec3e668034d" />

kpfleming commented on fastly/resource_fastly_ngwaf_virtualpatches.go:
```diff
@@ -36,24 +36,30 @@ func resourceFastlyNGWAFVirtualPatches() *schema.Resource {
 }
 
+func resourceFastlyNGWAFVirtualPatchCreate(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	return diag.Errorf("New Virtual patches cannot be created. Virtual patches must already exist in your Fastly NGWAF workspace. Use this resource to configure existing virtual patches only. Ensure the virtual_patch_id '%s' exists in workspace '%s' before applying.",
```
```suggestion
	return diag.Errorf("Virtual Patches cannot be created. Virtual Patches must already exist in your Fastly Next-Gen WAF workspace. Use this resource to configure existing Virtual Patches only. Ensure the virtual_patch_id '%s' exists in workspace '%s' before applying.",
```

kpfleming replied:
How do customers get Virtual Patches 'into' their workspaces?

rcaril replied:
VPs are pre-existing, they exist in every workspace no matter what. The customer only has the option to turn one off and change the mode.

kpfleming replied:
OK, we might need to improve the language here then, as it implies that the customer needs to do something to get the VP to exist in the workspace. I would just remove the second sentence and leave the others, and remove 'only'.

kpfleming commented on fastly/resource_fastly_ngwaf_virtualpatches.go:
```diff
@@ -64,7 +70,7 @@ func resourceFastlyNGWAFVirtualPatchRead(ctx context.Context, d *schema.Resource
 	log.Printf("[DEBUG] REFRESH: NGWAF virtual patch input: %#v", i)
 
-	virtualpatch, err := ws.Get(gofastly.NewContextForResourceID(ctx, d.Id()), conn, &i)
+	virtualpatch, err := ws.Get(gofastly.NewContextForResourceID(ctx, d.Get("workspace_id").(string)), conn, &i)
```
Please make the same change for the `Update` and `Delete` functions.

rcaril replied:
ooops

kpfleming reviewed (changes requested):
We still need `templates/data-sources/ngwaf_virtual_patches.md`, and some minor other changes, but otherwise this is in great shape.

rcaril committed:
+  Data Source Template & Create Error Verbiage Updates

rcaril committed:
Merge branch 'main' into CDTOOL-1108

rcaril committed:
Linting Fix

rcaril committed:
lint x2

kpfleming reviewed (approved):
Ship it!
